### PR TITLE
Add Cardano Helpers

### PR DESCRIPTION
**Pull request by tale-fb**

### CODE REVIEW

File: utils/ada_helpers.py [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+            if 'transaction' in utxo and 'block' in utxo['transaction'] and 'number' in utxo['transaction']['block'] \
+                    and 'txHash' in utxo and 'index' in utxo and 'address' in utxo and 'value' in utxo:
+                has_tokens = 'tokens' in utxo and (len(utxo['tokens']) > 0)
+                if has_tokens and native_only:
+                    continue
+
+                utxo_address = utxo['address']
+                utxo_tokens = []
+                for token in utxo['tokens']:
+                    if 'quantity' in token and 'asset' in token \
```
Comment (omereli) [inline utils/ada_helpers.py:123]: Please validate if has_tokens = false before,
if utxo['tokens'] doesn't exists we will throw

File: utils/ada_helpers.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+    def get_utxos(self, addresses: List[CardanoAddress], native_only: bool) -> List[CardanoUTxO]:
+        query = r'''query utxoSetForAddresses ( $addresses: [String]!) { utxos( where: { address: { _in: $addresses }} ) { address transaction { block { number } } tokens { asset { name ticker } quantity } txHash index value } }'''
+
+        addresses_str_arr = []
+        addresses_index_map = dict()
+        for address in addresses:
+            addresses_str_arr.append(address.address)
+            addresses_index_map[address.address] = address.derivation_index
+
+        result = self.__gql_server_caller(self.__gql_server_url, query, {'addresses': addresses_str_arr})
```
Comment (omereli) [inline utils/ada_helpers.py:107]: consider adding print to result

File: utils/ada_helpers.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+        if 'cardano' in result and 'tip' in result['cardano'] and 'slotNo' in result['cardano']['tip']:
+            return result['cardano']['tip']['slotNo']
+        else:
+            raise Exception(f'Bad response from GQL: {result}')
+
+    def submit_tx(self, signed_tx: bytes) -> str:
+        query = r'''mutation SendRawTx($rawtx: String!) {submitTransaction(transaction: $rawtx) {hash}}'''
+        result = self.__gql_server_caller(self.__gql_server_url, query, {'rawtx': signed_tx.hex()})
+        if 'submitTransaction' in result and 'hash' in result['submitTransaction']:
+            return result['submitTransaction']['hash']
```
Comment (omereli) [inline utils/ada_helpers.py:94]: consider printing result on success as well

File: utils/ada_helpers.py [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+
+        return max_index
+
+
+    def __get_balance_for_addresses(self, addresses: List[CardanoAddress]) -> CardanoBalanceMap:
+        balances_map = CardanoBalanceMap()
+        for utxo in self.__gql_client.get_utxos(addresses, True):
+            balances_map.increase_balance('ADA', utxo.native_amount)
+            for token in utxo.tokens:
+                balances_map.increase_balance(token.name, token.amount)
```
Comment (omereli) [inline utils/ada_helpers.py:344]: I think we shouldn't add it to the overall balance. What is the output in graphql if you query paymentAddresses? I think it's separate per asset.
However, as we don't support tokens yet, we shouldn't receive any utxo.tokens,
consider throwing unsupported for now.

File: utils/ada_helpers.py [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+        return requests.post(
+            gql_server_url,
+            headers={
+                'Content-Type': 'application/json'
+            },
+            json={
+                'query': str(queryStr),
+                'params': str(params).replace("\'", "\"") if params else None
+            }
+        ).json()
```
[the remark this answers is not in the document]
Comment (tale-fb) [reply utils/ada_helpers.py:157]: I'd like to keep this file without any prints. 
This makes a try-except block redundant (as the except clause would just print the exception's error).

File: utils/ada_helpers.py [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+        self.__fpub = fpub.encode('utf-8')
+        self.__account = account
+        self.__gql_client = gql_client
+        self.__mainnet = mainnet
+        self.__addresses = []
+
+        if mainnet:
+            self.__coin_type = ADA_COIN_TYPE
+        else:
+            self.__coin_type = 1
```
[the remark this answers is not in the document]
Comment (tale-fb) [reply utils/ada_helpers.py:175]: :+1:

File: utils/ada_helpers.py [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+            self.__min_address_index = min_address_index
+
+        if max_address_index is None:
+            self.__max_address_index = self.__find_max_address_index(DEFAULT_ADDRESS_GAP_POOL)
+        else:
+            self.__max_address_index = max_address_index
+
+        if self.__min_address_index < 0 or self.__max_address_index < self.__min_address_index:
+            raise Exception(f'Invalid min_address_index value of {self.__min_address_index} '
+                            f'(max: {self.__max_address_index})')
```
[the remark this answers is not in the document]
Comment (tale-fb) [reply utils/ada_helpers.py:189]: :+1:

File: utils/ada_helpers.py (hunk 1 above)
Comment (tale-fb) [reply utils/ada_helpers.py:123]: :+1:

File: utils/ada_helpers.py (hunk 2 above)
Comment (tale-fb) [reply utils/ada_helpers.py:107]: I'd like to keep this file without any prints.

File: utils/ada_helpers.py (hunk 3 above)
Comment (tale-fb) [reply utils/ada_helpers.py:94]: I'd like to keep this file without any prints.

File: utils/ada_helpers.py (hunk 4 above)
Comment (tale-fb) [reply utils/ada_helpers.py:344]: Not sure if I agree, but lets leave it for the next iteration. 
I don't think we should throw, we can allow this information to be retrieved.

Comment (omereli) [reply utils/ada_helpers.py:347]: Even when we do support tokes, it should be separated by asset, as it's not an ADA balance

**Pull request merged**